src/librustc_trait_selection/traits/error_reporting/mod.rs

```diff
@@ -284,18 +284,51 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                             .unwrap_or(false);
                         let is_from = format!("{}", trait_ref.print_only_trait_path())
                             .starts_with("std::convert::From<");
+                        let should_convert_option_to_result =
+                            format!("{}", trait_ref.print_only_trait_path())
+                                .starts_with("std::convert::From<std::option::NoneError");
+                        let should_convert_result_to_option = format!("{}", trait_ref)
+                            .starts_with("<std::option::NoneError as std::convert::From<");
                         let (message, note) = if is_try && is_from {
-                            (
-                                Some(format!(
+                            if should_convert_option_to_result {
+                                (
+                                    Some(format!(
+                                        "`?` couldn't convert the error to `{}`",
+                                        trait_ref.self_ty(),
+                                    )),
+                                    Some(
+                                        "the question mark operation (`?`) implicitly performs a \
+                                        conversion on the error value using the `From` trait. \
+                                        You might want to convert the Option into a Result using Option::ok_or"
+                                            .to_owned(),
+                                    ),
+                                )
```

Comment:
Instead of modifying the note, we could add a structured suggestion further down after `err` is created. It can be something like

```
err.span_suggestion_verbose(
    span.shrink_to_lo(),
    "consider converting the `Option<T>` into a `Result<T, _>" using `Option::ok_or` or `Option::ok_or_else", // Ideally we should get the type of `T` and present that
    ".ok_or_else(|_| /* error value */)".to_string(),
    Applicability::HasPlaceholders,
);
```

We can do this for the `Result` to `Option` case too.